src/compiler/crystal/syntax/parser.cr

```diff
@@ -997,7 +997,9 @@ module Crystal
           method = "[]"
         end
         location = @token.location
-        node_and_next_token Call.new(Global.new("$~").at(location), method, NumberLiteral.new(value.to_i))
+        index = value.to_i?
+        raise "Regex capture index #{value} doesn't fit in an Int32" unless index
```

Comment:
I don't recall the term "regex capture index". At https://crystal-lang.org/api/1.2.2/Regex.html they're called "special variables". I couldn't find any mention in the language specification (that's another TODO apparentyl).

Reply:
"special variables" isn't the relevant part from it; I'm referring to "capture groups" actually, which are definitely a thing.
This is "index of a regex capture group", and I considered the shortened form of that to be acceptable. Let me know what you think.

Reply:
Let me know if you see this as a blocking comment, because at the moment I'm not sure how to resolve this or if anything needs to be done.

Well, for one, regarding your direct suggestion, I don't think putting "special variable" would be an improvement here, it's not really about special variables in general.

Reply:
Sorry, I missed your reply.

I'm pretty sure this is about special variables in general. In stdlib, we use these special variables for regex capture groups. But the compiler is completely oblivious to that. It doesn't care *what* you use it for. You could employ special variables for completely different use cases. I the compiler error message should not assume a specific use case. THe example in spec `$99999999999999999999999?` really has nothing to do with regular expressions.

Reply:
Nah, they were done like that to simplify how they are used, but they aren't meant to be used for other purposes. I think it's fine to assume these are for regex. Otherwise the error message will be extremely confusing.

Reply:
> could employ special variables for completely different use cases

Right-- Now I see what you mean. The wording I have put isn't technically correct... Still though, "special variables" would be way too vague; after all, `$?` is also just a special variable. Do you have another alternative that we might agree on?

Reply:
What about "indexed special variable"?

Reply:
Mentioning "regex" part is IMO important to clarify the context.

Reply:
"indexed special variable" what?

> Index of indexed special variable 123 doesn't fit in an Int32

???

Anyway I don't like this, and seems we have 3 people saying that mentioning regex is good, so I don't think we need to stall this.

Reply:
`Indexed special variable $123 is invalid`

Reply:
This conveys much less information. Why is it invalid?

Reply:
And why _indexed_? Indexed mean `[0]`, and there's nothing like that anywhere in sight.

Reply:
@oprypin I don't think it matters much to provide more detail. It's very unlikely to happen and if it does, it should be reasonably clear that you just can't have a special variable with 10 digits. But feel free to add more information if you think it's relevant.
@Sija These special variables are a direct index into `$~`. Since this expansion happens here in the parser, you can even see that when you inspect the variable read (`p! $1` prints `$~[1] # => ...`).
I don't insist on "indexed" though. "special variable" would be fine by me. But I refuse to accept a newly constructed term that's too narrow to properly represent the concept.

Reply:
Then I suppose we'll stay with the compiler crash for now

Reply:
The thing is...
This is not even a special variable, it's an on-the-fly expansion into one. If we don't fully spell that out, saying "special variable" is just as wrong, if not more so.

Reply:
If we assume the context is regex, then the original error message is fine. I would add the `$` though, to make it more evident to what is it referring to. If later these variables/indexes are used for another purpose, we can rethink a better name. But this is an error that I won't expect to trigger very often to worry too much.

Reply:
> If later these variables/indexes are used for another purpose

There is simply no way to know *what* they are used for. Only in stdlib we use them exclusively for regex matchdata. But there's no restriction on assigning other meanings in user code. 

A conceivable alternative use could be to access the process arguments like in a shell:
```cr
def args
  $~ = [PROGRAM_NAME, *ARGV]
end
args
p! $1
```

Reply:
I doubt anyone besides you and me, and maybe a few others, know they can do `$~ = ...` in their code.

What I would do, if we really wanted to be strict about this, is to limit the types that can be assigned to `$~` and `$?` to `Regex::MatchData?` and `Process::Status?` respectively.

I don't want to encourage people to use them for anything else.

Reply:
That would require the compiler to know about these types. That's 👎

Reply:
Why? It's really easy to do.

Reply:
This is done.
Thanks for the non-conflicting good suggestion.